stubs/Werkzeug/werkzeug/datastructures.pyi

```diff
@@ -191,7 +191,10 @@ class Headers(object):
     def extend(self, iterable): ...
     def __delitem__(self, key: Any) -> None: ...
     def remove(self, key): ...
-    def pop(self, **kwargs): ...
+    @overload
+    def pop(self, key: str, default: Optional[str] = ...) -> str: ...
```

Comment:
Does `headers.pop("asdf", None)` really return `str`?

Reply:
No it return a `None` value, I guess it should be the type of default argument.

Reply:
Yes, that works.

Additionally, if you provide a non-`None` default, can the return value still be `None`? If not, it might be good to add a new overload just for that case.

Reply:
Basically, `pop` checks if the key is present, if not it will take the default value whatever it maybe. So, the `return` value will have the same type as `default` value type. So, non-`None` header will return non-`None` value if str key not found. How should I overload for this condition ?

Reply:
I think this should work:

```python
@overload
def pop(self, key: Optional[int] = ...) -> str: ...  # default is ignored, using it is an error
@overload
def pop(self, key: str) -> str: ...
@overload
def pop(self, key: str, default: str) -> str: ...
@overload
def pop(self, key: str, default: None) -> Optional[str]: ...
```

If the key is `None` or `int`, or not specified, we don't accept a `default` at all; it would be ignored anyway and that would be confusing. Otherwise, the key is a string, and it returns a string in all cases except when the default is `None`; then it might or might not return a string, depending on whether it finds the header or uses the default.

There are still some cases that this doesn't cover, such as `headers.pop("lol", 123)`, which can be `str` or `int`. But I don't think that comes up a lot in practice. If it does, we could do `def pop(..., default: _T) -> Union[str, _T]` where `_T` is a `TypeVar`.

Reply:
You are right about the last case, that case won't come up much often in real-time scenario. 
Following are the final 5 methods. I guess these cover every scenario as you suggested. Let me know if this is ok.
```python
@overload
def pop(self, key: Optional[int] = ...) -> str: ...  # default is ignored, using it is an error
@overload
def pop(self, key: str) -> str: ...
@overload
def pop(self, key: str, default: str) -> str: ...
@overload
def pop(self, key: str, default: None) -> Optional[str]: ...
@overload
def pop(self, key: Optional[int], default: Optional[Tuple[str, str]] = ...) -> Tuple[str, str]: ...
```

Reply:
You added an overload for `key: Optional[int]` that accepts default. I think it doesn't make sense.

Are we looking at the same source code? This is what I'm looking at. Notice how it doesn't do anything with the default if `key is None` or it's an integer:

```python
    def pop(self, key=None, default=_missing):
        """..."""
        if key is None:
            return self._list.pop()
        if isinstance(key, integer_types):
            return self._list.pop(key)
        ...
```

Reply:
Ohh my bad, I think the above 4 methods should suffice.